### PR TITLE
[CUDAGraph] add config `cudagraph_capture_sizes`

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Users may want CUDAGraph for certain sizes and fallback for other sizes.

As discussed in Issue #121968, we would like to use cudagraph for [batch size [1,2,3,...,16]](https://github.com/pytorch/pytorch/issues/121968#issuecomment-2259942345) and fallback for others. 

Another use case is [vllm](https://github.com/vllm-project/vllm/blob/main/vllm/compilation/cuda_piecewise_backend.py#L114-L119), where 67 batch sizes (i.e., [1,2,4,8,16,24,32,...,512]) are captured and all other sizes fallback.

This PR implements the feature with `torch._inductor.config.triton.cudagraph_capture_sizes`. When it is specified, we only capture cudagraph for these shapes. When it is None (by default), we capture cudagraph for all shapes.

Example:
```python
import torch

torch._inductor.config.triton.cudagraph_capture_sizes = [(2,3), (4,5), (6, 2), (7,3)]

def f(x):
    return x + 1

f = torch.compile(f, mode="reduce-overhead", dynamic=False)

def run(batch_size, seq_len, d):
    x = torch.randn((batch_size, seq_len, d), device="cuda")
    # Need to mark the dimension as dynamic. Automated-dynamic
    # may have some ux issues on matching `cudagraph_capture_sizes`
    # with the actual dynamic shapes, since there are specialization and
    # multiple dynamo graphs.
    torch._dynamo.mark_dynamic(x, 0)
    torch._dynamo.mark_dynamic(x, 1)
    for _ in range(3):
        f(x)

for i in range(2, 10):
    for j in range(2, 10):
        run(i, j, 8)

num_cudagraph = torch._inductor.cudagraph_trees.get_container(0).tree_manager.new_graph_id()
assert num_cudagraph.id == 4
```

cc @mcarilli @ezyang @eellison @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Do these checks apply only to the batch size? or also to seqlen? Or do full batch shapes need/can be specified?

**Comment by BoyuanFeng** [comment]
> Do these checks apply only to the batch size? or also to seqlen?

It applies to any dimension. Users need to `torch._dynamo.mark_dynamic(x, dynamic_dimension)` and make sure it matches `torch._inductor.config.triton.cudagraph_capture_sizes`. See above examples.

When user mark many dynamic dimensions, there might be an issue on matching the order of tuples (i.e., (2,3,4,8,10) vs (3,4,2,10,8)) between `torch._inductor.config.triton.cudagraph_capture_sizes` and dynamic shape order in the generated code. But it should be fine for existing use cases.

>  Or do full batch shapes need/can be specified?

No. Only need to mark necessary dimensions as dynamic and provide the shapes. All other dimension should be static.

**Comment by BoyuanFeng** [comment]
size 1 is always specialized, which makes specify cudagraph_capture_sizes with 1 a bit hard
```
import torch

def f(x):
    return x + 1

f = torch.compile(f)

x = torch.randn((1,5), device="cuda")
torch._dynamo.mark_dynamic(x, 0)
f(x)
```
gives
image

**Comment by galv** [comment]
There are two concerns that are worth discussing:

1. This doesn't do padding by default. I understand how this would complicate things. However, think about what will end up happening today. The user will have to copy data into a padded buffer, and then this padded buffer's contents would have to be copied again into the cuda graph's static buffer. These copies happen serially. Secondly, if you have multiple inputs, each copy will likely happen in a single cuda kernel per copy, serially. (torch.compile() can optimize this into  a single kernel that does SM-based copies on every input at once).

2. https://github.com/pytorch/pytorch/pull/156551/files#diff-3e1199aa4b118c07d8c38c05f1ced6d2c87b6b94f868fbc7596d27c06d5a8f9bR383 does not use the same memory pool as the cudagraph tree uses. We can assume that there is only one call to this torch.compile-emitted function at a single time (right? I'm actually not sure that that is true, thinking more on it. Someone could just run on two separate streams...). In that case, it should be safe to run this under: https://github.com/pytorch/pytorch/blob/c82a174ceab79f77ba18405dd263eb03692608fd/torch/_inductor/cudagraph_trees.py#L549-L573 This seems like it would obviously improve memory usage. The only concern I can think of is that you may need to copy out your outputs into a different buffer not owned by the memory pool.

File: test/inductor/test_cudagraph_trees.py [hunk 1]
```diff
@@ -3534,6 +3534,79 @@ def foobar(x, y):
             self.assertEqual(eager_out, compiled_out)
             self.assertEqual(self.get_manager().new_graph_id().id, 1)
 
+        def test_cudagraph_capture_sizes(self):
+            torch._inductor.config.triton.cudagraph_capture_sizes = [2, 5, 7]
```
Comment (bobrenjc93) [inline test/inductor/test_cudagraph_trees.py]: this should be a tuple or you should update the type definition?

File: test/inductor/test_cudagraph_trees.py [hunk 2]
```diff
@@ -3534,6 +3534,79 @@ def foobar(x, y):
...
+                for _ in range(3):
+                    f(x)
+
+            for i in range(1, 10):
+                run(i)
+
+            self.assertEqual(self.get_manager().new_graph_id().id, 3)
+
+        def test_cudagraph_capture_sizes1(self):
+            torch._inductor.config.triton.cudagraph_capture_sizes = [
```
Comment (bobrenjc93) [inline test/inductor/test_cudagraph_trees.py]: ditto

File: test/inductor/test_cudagraph_trees.py [hunk 3]
```diff
@@ -3534,6 +3534,79 @@ def foobar(x, y):
...
+                    f(x)
+
+            for i in range(2, 10):
+                for j in range(2, 10):
+                    run(i, j, 8)
+
+            self.assertEqual(self.get_manager().new_graph_id().id, 4)
+
+        def test_cudagraph_capture_sizes2(self):
+            torch._inductor.config.triton.cudagraph_capture_sizes = [
```
Comment (bobrenjc93) [inline test/inductor/test_cudagraph_trees.py]: ditto

File: torch/_inductor/cudagraph_trees.py [hunk 4]
```diff
@@ -346,6 +346,17 @@ def get_manager(
...
 
 
+def is_cudagraph_capture_sizes(int_key: Union[int, tuple[int, ...]]) -> bool:
+    """
+    Returns true if all dynamic shapes should be captured or the dynamic shape
+    int_key should be captured.
+    """
+    return (
+        config.triton.cudagraph_capture_sizes is None
+        or int_key in config.triton.cudagraph_capture_sizes
```
Comment (bobrenjc93) [inline torch/_inductor/cudagraph_trees.py:356]: hmmm if int_key is a tuple, are we sure this does what we want? I would have assumed this would return true

```
>>> (1,2) in (1,2,3)
False
```

Comment (bobrenjc93) [reply torch/_inductor/cudagraph_trees.py:356]: nvm I guess this is more representative of the use case

```
>>> (1,2) in ((1,2),3)
True
```

**Review by bobrenjc93 — APPROVED** [review]

**Comment by BoyuanFeng** [comment]
@galv thanks for the comments!

> This doesn't do padding by default.

Yes this does not pad. The use case is not turning on cudagraph by default (for now). The goal is to provide an api to specify a few sizes that users think it is important to cudagraph, and fallback to eager for all other sizes. This is how vllm currently uses cudagraph: [capturing 67 sizes and fallback for others](https://github.com/vllm-project/vllm/blob/main/vllm/compilation/cuda_piecewise_backend.py#L114-L119). vllm is also not doing padding.

The motivation is that a user may be happy to spend some startup time to record cudagraph, but wants to avoid a qps drop after that when observing a new shape.

> run with `_use_cuda_memory_pool_manager `

This would allocate output in the memory pool which is kind of like run_eager. My concerns are 1) we probably don't want to pay overhead of run_eager, since we will never record a cudagraph for this (fallback) size; 2) if we use `_use_cuda_memory_pool_manager` directly without `run_eager`, we need to manually deallocate the output tensor memory later.

Probably we should treat this fallback as other `skipping cudagraph due to [some reasons]`, which also run without `_use_cuda_memory_pool_manager`.

**Comment by eellison** [comment]
I agree with @galv that we should use the cudagraph memory pool for this run. we go through all the care of warming up cudagraph trees eager in cudagraph pool because we don't want to do a run of the model with the eager pool after allocation to cudagraph pool (which 2x memory use). This is a very similar use case to that.
